crates/holochain/src/conductor/conductor/agent_key_operations.rs

```diff
@@ -0,0 +1,152 @@
+use holochain_types::deepkey_roundtrip_backward;
+
+use super::*;
+
+/// The result type of an agent key revocation for an app.
+pub type RevokeAgentKeyForAppResult = HashMap<CellId, ConductorApiResult<()>>;
+
+impl Conductor {
+    /// Revoke an agent's key pair for all cells of an app.
+    ///
+    /// Writes a `Delete` action to the source chain of all cells of the app, which renders them read-only.
+    /// If DPKI is installed as conductor service, the agent key will be revoked there too and becomes
+    /// invalid.
+    pub async fn revoke_agent_key_for_app(
+        self: Arc<Self>,
+        agent_key: AgentPubKey,
+        app_id: InstalledAppId,
+    ) -> ConductorResult<RevokeAgentKeyForAppResult> {
+        // Disable app while revoking key
+        self.clone()
+            .disable_app(app_id.clone(), DisabledAppReason::DeletingAgentKey)
+            .await?;
+
+        // Revoke key in DPKI first, if installed, and then in cells' source chains.
+        // Call separate function so that in case a part of key revocation fails, the app is still enabled again.
+        let revocation_per_cell_results =
+            Conductor::revoke_agent_key_for_app_inner(self.clone(), agent_key, app_id.clone())
+                .await;
+
+        // Enable app again.
+        self.clone().enable_app(app_id.clone()).await?;
+
+        let deletion_per_cell_results = revocation_per_cell_results?;
+
+        // Publish 'Delete' actions of cells where successful.
+        // Triggering workflow is only possible when cells are enabled.
+        let publish_workflow_triggers = deletion_per_cell_results
+            .iter()
+            .filter(|(_, result)| result.is_ok())
+            .map({
+                |(cell_id, _)| {
+                    let conductor = self.clone();
+                    async move {
+                        let cell_triggers = conductor.get_cell_triggers(cell_id).await?;
+                        cell_triggers.publish_dht_ops.trigger(&"delete agent key");
+                        ConductorApiResult::Ok(())
+                    }
+                }
+            });
+        let _trigger_publish_results = futures::future::join_all(publish_workflow_triggers).await;
+
+        // Return cell ids with their agent key deletion result
+        Ok(deletion_per_cell_results)
+    }
+
+    /// Revoke agent key in Deepkey first, if installed, and then write a [`Delete`] of the key to the source chain.
+    async fn revoke_agent_key_for_app_inner(
```

Comment:
It seems like this creates a race condition. Isn't it the case that other agents are not going to be able to validate the Delete action, because the agent has already been revoked in DPKI? So shouldn't the DPKI revocation happen only after all Delete actions have been committed? Especially so if the revocation is partial and needs to be retried.

Reply:
> When you revoke it, it's revoked in DPKI and from then on you can't write more actions

This is the crux of what I'm saying. `Delete` is also an action, and if your agent key is revoked in deepkey, you shouldn't even be able to write that action. Unless we make that a special case, but I question the benefit of making that exception.

Reply:
This works quite nicely as desired now without creating exceptions. After revoking the agent key in DPKI, keys are deleted directly on the source chains. Zome calls committing entries will fail because they're sys validated. But you can retry the revocation call and the `Delete` action is directly committed to the source chain if it's not been deleted yet. Being the only way to delete keys makes this call a secure operation.

Reply:
I don't understand how this works, given that [we check DPKI agent validity for every op, including Delete](https://github.com/holochain/holochain/blob/e3930b67bf7b12c8cfbb4cde6a30ce0ee4512c91/crates/holochain/src/core/workflow/sys_validation_workflow.rs#L639-L645). Is there a test that a delete is properly gossiped and integrated by other nodes?

Reply:
This is my only concern, if this is resolved then I approve

Reply:
Do not approve, for I was mistaken! You pointed out the flaw in this. My thinking was simplistic, reductionist, I was a fool to think things so easy. Allow me to correct it and I shall request your review anew.

Reply:
@maackle It worked because sys validation is not run when writing to the source chain and flushing. I've captured this as a possible shortcoming in the workflow review issue.

However, it's true that sys validation was failing on other conductors when receiving the ops of a key deletion after the key had been revoked in DPKI.

The adaption to make it possible to delete keys on source chains after they've been revoked in DPKI is here https://github.com/holochain/holochain/pull/3930/files#diff-ecca2cd75dc5730737f8aa8e6f7915a0221290c348272bb8563bb9ba38da5e0fR426

There are two tests now to make sure this works:
- Revocation with 2 cells https://github.com/holochain/holochain/pull/3930/files#diff-14062bb2839544a82ec826510201978b91f795dd579db231ca46127fe28d09e6R493
- Recovery from partial revocation https://github.com/holochain/holochain/pull/3930/files#diff-14062bb2839544a82ec826510201978b91f795dd579db231ca46127fe28d09e6R620

Reply:
That's a big oversight that we weren't running sys validation during commits! It's an easy fix but glad you found it.